File: src/sharpen.ts
import { createInterface, type Interface } from 'node:readline'
import { runAgent } from './agents.ts'
import type { Agent } from './config.ts'
import { composeInEditor } from './editor.ts'
import { emit, type Hooks } from './hooks.ts'
import { log } from './log.ts'
import { sharpenPrompt, sharpenReviewPrompt } from './prompts.ts'

// Interactive sharpen step. Seeded with the raw intent, an agent interrogates it
// into a self-contained goal spec, exploring the repo itself. Each agent turn is
// a slow research pass, so it BATCHES its questions; the CLI then walks the human
// through them one at a time (no per-question latency) and sends the answers back
// as one turn. Returns the refined intent + verify, or null if the human cancels.

export type SharpenResult = { intent: string; verify: string | null }
type SharpenOpts = {
  root: string
  agent: Agent
  reviewer?: Agent
  hooks: Hooks
  intent: string
  verify: string | null
}

type Turn = { role: 'human' | 'agent' | 'reviewer'; text: string }

// Bound the conversation so a misbehaving agent can't loop forever.
const MAX_TURNS = 24

function transcript(turns: Turn[]): string {
  return turns.map((t) => `${t.role}: ${t.text}`).join('\n\n')
}

type Parsed = { ready: boolean; verify: string | null; spec: string; message: string }

// The agent signals completion with a `SPEC READY` line; until then everything
// is conversational. Parse out the spec + verify when present.
function parseSharpen(text: string): Parsed {
  const marker = /^SPEC READY\s*$/m.exec(text)
  if (!marker) {
    return { ready: false, verify: null, spec: '', message: text.trim() }
  }
  const message = text.slice(0, marker.index).trim()
  const after = text.slice(marker.index + marker[0].length)
  const vmatch = /^VERIFY:\s*(.*)$/m.exec(after)
  const rawVerify = vmatch?.[1]?.trim() ?? ''
  const verify = rawVerify && !/^none$/i.test(rawVerify) ? rawVerify : null
  const spec = (vmatch ? after.slice(vmatch.index + vmatch[0].length) : after).trim()
  return { ready: true, verify, spec, message }
}

type Question = { q: string; rec: string }
type QuestionReply = { kind: 'cancel' } | { kind: 'finish' } | { kind: 'answers'; text: string }

// When not ready, the agent batches its questions in a `QUESTIONS` block of
// `- <question> ||| <recommended answer>` lines (any grounding context goes
// before the block). Empty questions means it replied in plain prose instead.
function parseQuestions(text: string): { preamble: string; questions: Question[] } {
  const marker = /^QUESTIONS\s*$/m.exec(text)
  if (!marker) {
    return { preamble: '', questions: [] }
  }
  const preamble = text.slice(0, marker.index).trim()
  const questions: Question[] = []
  for (const line of text.slice(marker.index + marker[0].length).split('\n')) {
    const item = /^\s*-\s+(.+)$/.exec(line)
    if (!item?.[1]) {
      continue
    }
    const [q, rec] = item[1].split('|||')
    if (q?.trim()) {
      questions.push({ q: q.trim(), rec: rec?.trim() ?? '' })
    }
  }
  return { preamble, questions }
}

type Review =
  | { kind: 'pass' }
  | { kind: 'revise'; message: string }
  | { kind: 'questions'; preamble: string; questions: Question[] }

function firstLine(text: string): string {
  return (
    text
      .split('\n')
      .map((line) => line.trim())
      .find(Boolean) ?? ''
  )
}

function parseReview(text: string): Review {
  const marker = firstLine(text)
  if (marker === 'SHARPEN: PASS') {
    return { kind: 'pass' }
  }
  if (marker === 'SHARPEN: REVISE') {
    const start = text.indexOf(marker) + marker.length
    const message = text.slice(start).trim()
    return {
      kind: 'revise',
      message: message || 'Tighten the spec before showing it to the human.',
    }
  }
  const { preamble, questions } = parseQuestions(text)
  if (questions.length > 0) {
    return { kind: 'questions', preamble, questions }
  }
  return {
    kind: 'revise',
    message: text.trim() || 'Tighten the spec before showing it to the human.',
  }
}

function ask(rl: Interface, question: string): Promise<string> {
  return new Promise((resolve) => rl.question(question, resolve))
}

type Reply = { kind: 'cancel' } | { kind: 'done' } | { kind: 'text'; text: string }

// One conversational reply turn: an empty line (or /done) finishes, a line of text
// keeps refining, /edit composes a multi-line reply in $EDITOR, /cancel aborts.
// An empty /edit re-prompts without re-running the agent.
async function readReply(rl: Interface): Promise<Reply> {
  while (true) {
    const input = (await ask(rl, 'you> ')).trim()
    if (input === '/cancel') {
      return { kind: 'cancel' }
    }
    if (input === '' || input === '/done') {
      return { kind: 'done' }
    }
    if (input === '/edit') {
      const edited = await composeInEditor()
      if (!edited) {
        log.info('  (nothing entered)')
        continue
      }
      log.log(`\n${renderAgent(edited)}\n`)
      return { kind: 'text', text: edited }
    }
    return { kind: 'text', text: input }
  }
}

async function readQuestionAnswers(
  opts: SharpenOpts,
  rl: Interface,
  questions: Question[]
): Promise<QuestionReply> {
  const answered: string[] = []
  for (let qi = 0; qi < questions.length; qi++) {
    const { q, rec } = questions[qi] ?? { q: '', rec: '' }
    log.log(`\n${BOLD}(${qi + 1}/${questions.length})${RESET} ${styleLine(q)}`)
    if (rec) {
      log.info(`  recommend: ${rec}`)
    }
    const input = (await waitingForInput(opts, () => ask(rl, 'you> '))).trim()
    if (input === '/cancel') {
      return { kind: 'cancel' }
    }
    if (input === '/done') {
      return { kind: 'finish' }
    }
    if (input === '/skip') {
      answered.push(`Q: ${q}\nA: (skipped)`)
      continue
    }
    answered.push(`Q: ${q}\nA: ${input || rec || '(no preference)'}`)
  }
  return { kind: 'answers', text: answered.join('\n\n') }
}

// Drive lifecycle hooks for the interactive sharpen step: active stage while the agent
// computes, and attention while factory is actually waiting at a human prompt.
async function thinking<T>(opts: SharpenOpts, work: () => Promise<T>): Promise<T> {
  await emit(opts.root, opts.hooks, 'attention', { state: 'none' })
  await emit(opts.root, opts.hooks, 'stage.change', { stage: 'sharpen', active: true })
  try {
    return await work()
  } finally {
    await emit(opts.root, opts.hooks, 'stage.change', { stage: 'sharpen', active: false })
  }
}

async function waitingForInput<T>(opts: SharpenOpts, wait: () => Promise<T>): Promise<T> {
  await emit(opts.root, opts.hooks, 'attention', { state: 'needs-input' })
  try {
    return await wait()
  } finally {
    await emit(opts.root, opts.hooks, 'attention', { state: 'none' })
  }
}

function fmtTok(n: number): string {
  return n >= 1000 ? `${(n / 1000).toFixed(1)}k` : `${n}`
}

const DIM = '\x1b[2m'
const BOLD = '\x1b[1m'
const CYAN = '\x1b[36m'
const GREEN = '\x1b[32m'
const RESET = '\x1b[0m'

// Inline styling for one line of the agent's markdown: bold a leading "Label:"
// (Recommendation:/Why:/headings), colorize `code` spans (dropping the noisy
// backticks), and render **bold**.
function styleLine(line: string): string {
  return line
    .replace(/^(\s*)([A-Z][A-Za-z ]{1,38}):/, (_m, sp, label) => `${sp}${BOLD}${label}:${RESET}`)
    .replace(/`([^`]+)`/g, `${CYAN}$1${RESET}`)
    .replace(/\*\*([^*]+)\*\*/g, `${BOLD}$1${RESET}`)
}

// Frame an agent turn with a dim left gutter so a multi-paragraph message reads
// as one block, visually distinct from your left-aligned `you>` input.
function renderAgent(text: string): string {
  return text
    .split('\n')
    .map((line) => `${DIM}│${RESET} ${styleLine(line)}`)
    .join('\n')
}

function showSpec(spec: string, verify: string | null): void {
  log.log(`\n${GREEN}${BOLD}┌─ proposed spec ────────────────────────────${RESET}`)
  log.log(renderAgent(spec))
  if (verify) {
    log.log(`${DIM}│${RESET} ${BOLD}verify:${RESET} ${CYAN}${verify}${RESET}`)
  }
  log.log(`${GREEN}${BOLD}└────────────────────────────────────────────${RESET}`)
  log.info('Enter to queue this · or type a reply to keep refining')
}

async function reviewSpec(
  opts: SharpenOpts,
  turns: Turn[],
  spec: SharpenResult,
  tally: { inTok: number; outTok: number }
): Promise<Review> {
  log.info('  …checking spec')
  const { text, usage } = await thinking(opts, () =>
    runAgent(opts.reviewer ?? opts.agent, {
      root: opts.root,
      prompt: sharpenReviewPrompt(transcript(turns), spec.intent, spec.verify),
      access: 'read',
    })
  )
  tally.inTok += usage.inputTokens
  tally.outTok += usage.outputTokens
  return parseReview(text)
}

export async function sharpen(opts: SharpenOpts): Promise<SharpenResult | null> {
  log.step('sharpen — clarifying intent into a spec')
  log.info(
    '  Enter accepts a recommendation / finishes · /skip a question · /edit for a long reply · /cancel abort'
  )
  await emit(opts.root, opts.hooks, 'stage.change', { stage: 'sharpen', active: false })
  const turns: Turn[] = [{ role: 'human', text: opts.intent }]
  let proposed: SharpenResult | null = null
  const tally = { inTok: 0, outTok: 0 }
  const rl = createInterface({ input: process.stdin, output: process.stdout })
  try {
    for (let i = 0; i < MAX_TURNS; i++) {
      log.info('  …thinking')
      const { text, usage } = await thinking(opts, () =>
        runAgent(opts.agent, {
          root: opts.root,
          prompt: sharpenPrompt(transcript(turns), false),
          access: 'read',
        })
      )
      tally.inTok += usage.inputTokens
      tally.outTok += usage.outputTokens
      turns.push({ role: 'agent', text })
      const parsed = parseSharpen(text)

      // A batch of questions → walk them one at a time (no per-question latency),
      // then send all the answers back to the agent as a single turn.
      if (!parsed.ready) {
        const { preamble, questions } = parseQuestions(text)
        if (questions.length > 0) {
          if (preamble) {
            log.log(`\n${renderAgent(preamble)}`)
          }
          const reply = await readQuestionAnswers(opts, rl, questions)
          if (reply.kind === 'cancel') {
            return null
          }
          if (reply.kind === 'finish') {
            return proposed ?? (await finalizeReviewed(opts, turns, tally))
          }
          turns.push({ role: 'human', text: reply.text })
          continue
        }
      }

      // Otherwise: show the spec (if ready) or a plain prose reply, then take one
      // conversational reply (line, /edit, /done, /cancel).
      if (parsed.ready) {
        if (parsed.message) {
          log.log(`\n${renderAgent(parsed.message)}\n`)
        }
        const candidate = { intent: parsed.spec, verify: parsed.verify ?? opts.verify }
        const review = await reviewSpec(opts, turns, candidate, tally)
        if (review.kind === 'revise') {
          turns.push({
            role: 'reviewer',
            text:
              'Internal spec review asked for a revision before showing this to the human:' +
              `\n\n${review.message}`,
          })
          continue
        }
        if (review.kind === 'questions') {
          if (review.preamble) {
            log.log(`\n${renderAgent(review.preamble)}`)
          }
          const reply = await readQuestionAnswers(opts, rl, review.questions)
          if (reply.kind === 'cancel') {
            return null
          }
          if (reply.kind === 'finish') {
            turns.push({
              role: 'human',
              text: 'Use the spec reviewer recommended answers for those unresolved decisions.',
            })
            continue
          }
          turns.push({
            role: 'human',
            text: `Spec review identified unresolved human decisions:\n\n${reply.text}`,
          })
          continue
        }
        proposed = candidate
        showSpec(parsed.spec, proposed.verify)
      } else if (parsed.message) {
        log.log(`\n${renderAgent(parsed.message)}\n`)
      } else {
        log.info('(no response — type a reply to retry, or Enter to finish)')
      }
      const r = await waitingForInput(opts, () => readReply(rl))
      if (r.kind === 'cancel') {
        return null
      }
      if (r.kind === 'done') {
        return proposed ?? (await finalizeReviewed(opts, turns, tally))
      }
      turns.push({ role: 'human', text: r.text })
    }
    log.warn('sharpen: reached the turn limit — finalizing')
    return proposed ?? (await finalizeReviewed(opts, turns, tally))
  } finally {
    rl.close()
    await emit(opts.root, opts.hooks, 'attention', { state: 'none' })
    await emit(opts.root, opts.hooks, 'stage.change', { stage: '', active: false })
    if (tally.inTok || tally.outTok) {
      log.info(`  tokens: ${fmtTok(tally.inTok)} in → ${fmtTok(tally.outTok)} out`)
    }
  }
}

async function finalizeReviewed(
  opts: SharpenOpts,
  turns: Turn[],
  tally: { inTok: number; outTok: number }
): Promise<SharpenResult> {
  const candidate = await finalize(opts, turns, tally)
  const review = await reviewSpec(opts, turns, candidate, tally)
  if (review.kind === 'pass') {
    return candidate
  }
  if (review.kind === 'questions') {
    turns.push({
      role: 'human',
      text:
        'Use the spec reviewer recommended answers for these unresolved decisions:\n\n' +
        review.questions.map((q) => `Q: ${q.q}\nA: ${q.rec || '(no preference)'}`).join('\n\n'),
    })
    return await finalize(opts, turns, tally)
  }
  turns.push({
    role: 'reviewer',
    text: `Internal spec review asked for a final revision before queueing:\n\n${review.message}`,
  })
  return await finalize(opts, turns, tally)
}

// Force the agent to synthesize the spec from the conversation so far. Falls back
// to the raw intent if it can't — finishing should never lose the human's input.
async function finalize(
  opts: SharpenOpts,
  turns: Turn[],
  tally: { inTok: number; outTok: number }
): Promise<SharpenResult> {
  log.info('  …finalizing the spec')
  const { text, usage } = await thinking(opts, () =>
    runAgent(opts.agent, {
      root: opts.root,
      prompt: sharpenPrompt(transcript(turns), true),
      access: 'read',
    })
  )
  tally.inTok += usage.inputTokens
  tally.outTok += usage.outputTokens
  const parsed = parseSharpen(text)
  if (!parsed.ready || !parsed.spec) {
    log.warn('sharpen: could not synthesize a spec — keeping your original intent')
    return { intent: opts.intent, verify: opts.verify }
  }
  return { intent: parsed.spec, verify: parsed.verify ?? opts.verify }
}

export type GrillResult = SharpenResult
export const grill = sharpen

File: src/sharpen.rs
use regex::Regex;

pub const DIM: &str = "\x1b[2m";
pub const BOLD: &str = "\x1b[1m";
pub const CYAN: &str = "\x1b[36m";
pub const GREEN: &str = "\x1b[32m";
pub const RESET: &str = "\x1b[0m";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SharpenParsed {
    pub ready: bool,
    pub verify: Option<String>,
    pub spec: String,
    pub message: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Question {
    pub q: String,
    pub rec: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedQuestions {
    pub preamble: String,
    pub questions: Vec<Question>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Review {
    Pass,
    Revise {
        message: String,
    },
    Questions {
        preamble: String,
        questions: Vec<Question>,
    },
}

pub fn parse_sharpen(text: &str) -> SharpenParsed {
    let marker = Regex::new(r"(?m)^SPEC READY[ \t\r]*$").expect("valid regex");
    let Some(marker_match) = marker.find(text) else {
        return SharpenParsed {
            ready: false,
            verify: None,
            spec: String::new(),
            message: text.trim().to_string(),
        };
    };

    let message = text[..marker_match.start()].trim().to_string();
    let after = &text[marker_match.end()..];
    let verify_re = Regex::new(r"(?m)^VERIFY:[ \t\r]*(.*)$").expect("valid regex");
    let verify_match = verify_re.captures(after);
    let raw_verify = verify_match
        .as_ref()
        .and_then(|captures| captures.get(1))
        .map(|value| value.as_str().trim())
        .unwrap_or_default();
    let verify = if raw_verify.is_empty() || raw_verify.eq_ignore_ascii_case("none") {
        None
    } else {
        Some(raw_verify.to_string())
    };
    let spec = verify_match
        .as_ref()
        .and_then(|captures| captures.get(0))
        .map(|matched| after[matched.end()..].trim().to_string())
        .unwrap_or_else(|| after.trim().to_string());

    SharpenParsed {
        ready: true,
        verify,
        spec,
        message,
    }
}

pub fn parse_questions(text: &str) -> ParsedQuestions {
    let marker = Regex::new(r"(?m)^QUESTIONS[ \t\r]*$").expect("valid regex");
    let Some(marker_match) = marker.find(text) else {
        return ParsedQuestions {
            preamble: String::new(),
            questions: Vec::new(),
        };
    };

    let preamble = text[..marker_match.start()].trim().to_string();
    let item = Regex::new(r"^\s*-\s+(.+)$").expect("valid regex");
    let mut questions = Vec::new();
    for line in text[marker_match.end()..].lines() {
        let Some(captures) = item.captures(line) else {
            continue;
        };
        let Some(raw) = captures.get(1).map(|value| value.as_str()) else {
            continue;
        };
        let mut parts = raw.split("|||");
        let q = parts.next().map(str::trim).unwrap_or_default();
        let rec = parts.next().map(str::trim).unwrap_or_default();
        if !q.is_empty() {
            questions.push(Question {
                q: q.to_string(),
                rec: rec.to_string(),
            });
        }
    }

    ParsedQuestions {
        preamble,
        questions,
    }
}

pub fn parse_review(text: &str) -> Review {
    let marker = first_line(text);
    if marker == "SHARPEN: PASS" {
        return Review::Pass;
    }
    if marker == "SHARPEN: REVISE" {
        let start = text
            .find(&marker)
            .map(|idx| idx + marker.len())
            .unwrap_or(marker.len());
        let message = text[start..].trim();
        return Review::Revise {
            message: fallback_review_message(message),
        };
    }

    let parsed = parse_questions(text);
    if !parsed.questions.is_empty() {
        return Review::Questions {
            preamble: parsed.preamble,
            questions: parsed.questions,
        };
    }

    Review::Revise {
        message: fallback_review_message(text.trim()),
    }
}

pub fn format_questions(preamble: &str, questions: &[Question]) -> String {
    let mut blocks = Vec::new();
    let preamble = preamble.trim();
    if !preamble.is_empty() {
        blocks.push(preamble.to_string());
    }
    blocks.extend(questions.iter().map(|question| {
        if question.rec.is_empty() {
            format!("- {}", question.q)
        } else {
            format!("- {}\n  Recommended: {}", question.q, question.rec)
        }
    }));
    blocks.join("\n\n")
}

pub fn parse_formatted_questions(text: &str) -> ParsedQuestions {
    let question_re = Regex::new(r"^\s*-\s+(.+?)\s*$").expect("valid regex");
    let recommendation_re = Regex::new(r"^\s+Recommended:\s*(.*)$").expect("valid regex");
    let mut preamble = Vec::new();
    let mut questions = Vec::new();

    for block in Regex::new(r"\n{2,}")
        .expect("valid regex")
        .split(text.trim())
    {
        let lines: Vec<&str> = block.split('\n').collect();
        let question = lines
            .first()
            .and_then(|line| question_re.captures(line))
            .and_then(|captures| captures.get(1))
            .map(|value| value.as_str().trim().to_string());
        let recommendation = lines
            .get(1)
            .and_then(|line| recommendation_re.captures(line))
            .and_then(|captures| captures.get(1))
            .map(|value| value.as_str());

        let Some(q) = question else {
            preamble.push(block.to_string());
            continue;
        };
        let Some(first_rec_line) = recommendation else {
            preamble.push(block.to_string());
            continue;
        };

        let mut rec_lines = vec![first_rec_line];
        rec_lines.extend(lines.iter().skip(2).copied());
        questions.push(Question {
            q,
            rec: rec_lines.join("\n").trim().to_string(),
        });
    }

    ParsedQuestions {
        preamble: preamble.join("\n\n").trim().to_string(),
        questions,
    }
}

pub fn style_sharpen_markdown_line(line: &str) -> String {
    let heading = Regex::new(r"^(\s{0,3})(#{2,3})\s+(.+?)\s*#*\s*$").expect("valid regex");
    if let Some(captures) = heading.captures(line) {
        let sp = captures
            .get(1)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let marks = captures
            .get(2)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let text = captures
            .get(3)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let color = if marks == "##" { GREEN } else { CYAN };
        let heading_style = format!("{color}{BOLD}");
        return format!(
            "{sp}{heading_style}{}{RESET}",
            style_inline_markdown(text, &heading_style)
        );
    }

    let bullet = Regex::new(r"^(\s*)([-*]|\d+[.)])(\s+)(.+)$").expect("valid regex");
    if let Some(captures) = bullet.captures(line) {
        let sp = captures
            .get(1)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let marker = captures
            .get(2)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let gap = captures
            .get(3)
            .map(|value| value.as_str())
            .unwrap_or_default();
        let text = captures
            .get(4)
            .map(|value| value.as_str())
            .unwrap_or_default();
        return format!(
            "{sp}{DIM}{marker}{RESET}{gap}{}",
            style_inline_markdown(text, "")
        );
    }

    style_inline_markdown(line, "")
}

pub fn render_agent_markdown(text: &str) -> String {
    text.lines()
        .map(|line| format!("{DIM}|{RESET} {}", style_sharpen_markdown_line(line)))
        .collect::<Vec<_>>()
        .join("\n")
}

fn first_line(text: &str) -> String {
    text.lines()
        .map(str::trim)
        .find(|line| !line.is_empty())
        .unwrap_or_default()
        .to_string()
}

fn fallback_review_message(message: &str) -> String {
    if message.is_empty() {
        "Tighten the spec before showing it to the human.".to_string()
    } else {
        message.to_string()
    }
}

fn style_inline_markdown(line: &str, restore: &str) -> String {
    let label = Regex::new(r"^(\s*)([A-Z][A-Za-z ]{1,38}):").expect("valid regex");
    let code = Regex::new(r"`([^`]+)`").expect("valid regex");
    let bold = Regex::new(r"\*\*([^*]+)\*\*").expect("valid regex");

    let styled = label
        .replace(line, |captures: &regex::Captures<'_>| {
            format!("{}{}{}:{RESET}{restore}", &captures[1], BOLD, &captures[2])
        })
        .to_string();
    let styled = code
        .replace_all(&styled, |captures: &regex::Captures<'_>| {
            format!("{CYAN}{}{RESET}{restore}", &captures[1])
        })
        .to_string();
    bold.replace_all(&styled, |captures: &regex::Captures<'_>| {
        format!("{BOLD}{}{RESET}{restore}", &captures[1])
    })
    .to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn strip_ansi(text: &str) -> String {
        Regex::new(&format!("{}\\[[0-9;]*m", char::from(27)))
            .expect("valid regex")
            .replace_all(text, "")
            .to_string()
    }

    #[test]
    fn parses_ready_sharpen_output() {
        assert_eq!(
            parse_sharpen("context\nSPEC READY\nVERIFY: bun test\n\n## Goal\nShip it."),
            SharpenParsed {
                ready: true,
                verify: Some("bun test".to_string()),
                spec: "## Goal\nShip it.".to_string(),
                message: "context".to_string(),
            }
        );
        assert_eq!(parse_sharpen("SPEC READY\nVERIFY: none\nBody").verify, None);
    }

    #[test]
    fn unready_sharpen_output_is_message() {
        assert_eq!(
            parse_sharpen("Still investigating."),
            SharpenParsed {
                ready: false,
                verify: None,
                spec: String::new(),
                message: "Still investigating.".to_string(),
            }
        );
    }

    #[test]
    fn parses_batched_questions() {
        assert_eq!(
            parse_questions(
                "Grounding.\nQUESTIONS\n- Which scope? ||| Keep it narrow.\n- Any deadline?"
            ),
            ParsedQuestions {
                preamble: "Grounding.".to_string(),
                questions: vec![
                    Question {
                        q: "Which scope?".to_string(),
                        rec: "Keep it narrow.".to_string(),
                    },
                    Question {
                        q: "Any deadline?".to_string(),
                        rec: String::new(),
                    },
                ],
            }
        );
    }

    #[test]
    fn parses_review_markers_and_fallbacks() {
        assert_eq!(parse_review("\nSHARPEN: PASS\n"), Review::Pass);
        assert_eq!(
            parse_review("SHARPEN: REVISE\nAdd acceptance criteria."),
            Review::Revise {
                message: "Add acceptance criteria.".to_string(),
            }
        );
        assert_eq!(
            parse_review("QUESTIONS\n- Choose behavior? ||| Preserve current."),
            Review::Questions {
                preamble: String::new(),
                questions: vec![Question {
                    q: "Choose behavior?".to_string(),
                    rec: "Preserve current.".to_string(),
                }],
            }
        );
        assert_eq!(
            parse_review(""),
            Review::Revise {
                message: "Tighten the spec before showing it to the human.".to_string(),
            }
        );
    }

    #[test]
    fn formatted_questions_round_trip_recommendations() {
        let text = format_questions(
            "Grounding context.",
            &[
                Question {
                    q: "Should auto-upgrade skip source runs?".to_string(),
                    rec: "Yes, only installed binaries.".to_string(),
                },
                Question {
                    q: "Which commands are eligible?".to_string(),
                    rec: "Interactive normal commands only.".to_string(),
                },
            ],
        );

        assert_eq!(
            parse_formatted_questions(&text),
            ParsedQuestions {
                preamble: "Grounding context.".to_string(),
                questions: vec![
                    Question {
                        q: "Should auto-upgrade skip source runs?".to_string(),
                        rec: "Yes, only installed binaries.".to_string(),
                    },
                    Question {
                        q: "Which commands are eligible?".to_string(),
                        rec: "Interactive normal commands only.".to_string(),
                    },
                ],
            }
        );
    }

    #[test]
    fn formatted_question_parser_ignores_unrecommended_preamble_bullets() {
        let text = [
            "- Existing fact without a recommendation.",
            "",
            "- Which scope should apply?",
            "  Recommended: Keep it narrow.",
        ]
        .join("\n");

        assert_eq!(
            parse_formatted_questions(&text),
            ParsedQuestions {
                preamble: "- Existing fact without a recommendation.".to_string(),
                questions: vec![Question {
                    q: "Which scope should apply?".to_string(),
                    rec: "Keep it narrow.".to_string(),
                }],
            }
        );
    }

    #[test]
    fn renders_spec_headings_without_visible_markers() {
        let heading = style_sharpen_markdown_line("## Problem");

        assert!(heading.contains("\x1b[32m\x1b[1m"));
        assert_eq!(strip_ansi(&heading), "Problem");
        assert_eq!(
            strip_ansi(&style_sharpen_markdown_line("### Rejected Alternatives")),
            "Rejected Alternatives"
        );
    }

    #[test]
    fn renders_inline_emphasis_and_code_as_display_text() {
        let line = style_sharpen_markdown_line("Use **small** `terminal` markdown.");

        assert!(line.contains("\x1b[1msmall\x1b[0m"));
        assert!(line.contains("\x1b[36mterminal\x1b[0m"));
        assert_eq!(strip_ansi(&line), "Use small terminal markdown.");
    }

    #[test]
    fn keeps_list_structure_and_unsupported_markdown_readable() {
        let bullet = style_sharpen_markdown_line("- Use `bun test`");

        assert!(bullet.contains("\x1b[2m-\x1b[0m"));
        assert_eq!(strip_ansi(&bullet), "- Use bun test");
        assert_eq!(
            strip_ansi(&style_sharpen_markdown_line(
                "1. Read [README](./README.md)"
            )),
            "1. Read [README](./README.md)"
        );
    }
}
